## Kaggle Competition - House Prices: Advanced Regression Techniques
# Linear Regression Exploration

In [1]:
import sys
import os
sys.path.append( os.path.abspath( os.path.join(os.getcwd(), ".." ))) 
from src.utils import reset_root_dir
reset_root_dir()

'/Users/jamie/Dropbox/Programming/Kaggle/kaggle-house-prices'

In [2]:
import pandas as pd
from pandas import DataFrame, Series
import numpy as np
import math
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

from src.utils.Charts import Charts
from src.models.LinearRegressionModel import LinearRegressionModel

## Baseline Linear Regression

The basic workflow for linear regression is: 
- split the dataset into test and training datasets
- create a 2 dimentional dataframe of X input features and a 1 dimentional vector of known Y outputs
- generate artifical features as required
- fit a linear model
- score and output predictions against the test dataset

As a naive baseline, we can use the raw data from all the numeric columns in the dataset as X to predict the Y price. Non-numeric fields will cause linear regression to crash. 

In [20]:
linearRegressionModel = LinearRegressionModel()
linearRegressionModel.data["X_train"].head(3)

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  GarageArea  WoodDeckSF  \
0          2003       196.0         706  ...         548           0   
1          1976         0.0         978  ...         460         298   
2          2002       162.0         486  ...         608           0   

   OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
0           61              0          0            0         0        0   
1            0              0          0            0         0        0   
2           42              0          0            0         0        0   

   MoSold  YrSold  
0       2    2008  
1       5    2007  
2       9    2008  

[3 rows x 37 columns]

In [21]:
linearRegressionModel.data["Y_train"].head(3)

0    208500
1    181500
2    223500
Name: SalePrice, dtype: int64

In [15]:
linearRegressionModel.execute()

{'class': 'LinearRegressionModel',
 'filename': './data/submissions/BaselineLinearRegression.csv',
 'scores': {'LinearRegression': 0.8164275651873067}}

This gives us a local R^2 coefficient score of 0.816; with 0 being perfect prediction.

## Submit to Kaggle
```
$ kaggle competitions submit -c house-prices-advanced-regression-techniques -f data/submissions/LinearRegressionModel.csv -m "LinearRegressionModel.py - raw numeric fields
```

Your (Kaggle) submission scored 0.43452, which is an improvement of your previous score of 0.74279. Great job! 

Kaggle Rank 4079 / 4339